In [76]:
import pandas as pd
import numpy as np

df = pd.read_csv('titanic.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [77]:
df[df['Embarked'] == 'C'].groupby('Sex')['Age'].mean()

Sex
female    28.344262
male      32.998841
Name: Age, dtype: float64

In [78]:
df[df['Embarked'] == 'C']['Age'].mean()

30.81476923076923

In [79]:
(28.344262 + 32.998841)/2

30.6715515

# Question: Why we got 30.81 for average age of passengers?

hint: determine how many non Nan values exists for the age of female passengers in C

hint: determine how many not Nan values exists for the age of male passengers in C

In [80]:
df[(df['Embarked'] == 'C') & (df['Sex'] == 'female')].describe()

PassengerId   Survived     Pclass        Age      SibSp      Parch  \
count    73.000000  73.000000  73.000000  61.000000  73.000000  73.000000   
mean    437.315068   0.876712   1.726027  28.344262   0.547945   0.493151   
std     257.119021   0.331042   0.916822  15.770830   0.624524   0.766001   
min       2.000000   0.000000   1.000000   0.750000   0.000000   0.000000   
25%     257.000000   1.000000   1.000000  17.000000   0.000000   0.000000   
50%     381.000000   1.000000   1.000000  24.000000   0.000000   0.000000   
75%     642.000000   1.000000   3.000000  40.000000   1.000000   1.000000   
max     880.000000   1.000000   3.000000  60.000000   2.000000   3.000000   

             Fare  
count   73.000000  
mean    75.169805  
std     83.574380  
min      7.225000  
25%     15.245800  
50%     56.929200  
75%     91.079200  
max    512.329200

In [81]:
df[(df['Embarked'] == 'C') & (df['Sex'] == 'male')].describe()

PassengerId   Survived     Pclass        Age      SibSp      Parch  \
count    95.000000  95.000000  95.000000  69.000000  95.000000  95.000000   
mean    451.536842   0.305263   2.010526  32.998841   0.263158   0.263158   
std     262.426679   0.462962   0.950865  14.905440   0.466096   0.549865   
min      27.000000   0.000000   1.000000   0.420000   0.000000   0.000000   
25%     206.000000   0.000000   1.000000  24.000000   0.000000   0.000000   
50%     496.000000   0.000000   2.000000  30.000000   0.000000   0.000000   
75%     661.000000   1.000000   3.000000  40.000000   0.500000   0.000000   
max     890.000000   1.000000   3.000000  71.000000   2.000000   2.000000   

             Fare  
count   95.000000  
mean    48.262109  
std     82.715093  
min      4.012500  
25%      7.229200  
50%     24.000000  
75%     52.472950  
max    512.329200

# Solution: 

In [98]:
# determine how many non Nan values exists for the age of female passengers in C

df[(df['Embarked'] == 'C') & (df['Sex'] == 'female')]['Age'].isna().sum()

12

In [100]:
len(df[(df['Embarked'] == 'C') & (df['Sex'] == 'female')])

73

In [111]:
print('Female: Non Nan values in C =  ' + str(73 - 12))

Female: Non Nan values in C =  61


In [99]:
#determine how many not Nan values exists for the age of male passengers in C

df[(df['Embarked'] == 'C') & (df['Sex'] == 'male')]['Age'].isna().sum()

26

In [108]:
len(df[(df['Embarked'] == 'C') & (df['Sex'] == 'male')])

95

In [110]:
print('Male: Non Nan values in C =  ' + str(95-26))

Male: Non Nan values in C =  69


In [112]:
print('The number of nan value for the age of female passengers in C is:')
print(df[(df['Embarked'] == 'C') & (df['Sex'] == 'female')]['Age'].isna().sum())
print('The total size of dataset for female passengers in C is:')
len(df[(df['Embarked'] == 'C') & (df['Sex'] == 'female')])


print('The number of nan value for the age of male passengers in C is:')
print(df[(df['Embarked'] == 'C') & (df['Sex'] == 'male')]['Age'].isna().sum())
print('The total size of dataset for male passengers in C is:')
len(df[(df['Embarked'] == 'C') & (df['Sex'] == 'male')])

The number of nan value for the age of female passengers in C is:
12
The total size of dataset for female passengers in C is:
The number of nan value for the age of male passengers in C is:
26
The total size of dataset for male passengers in C is:


95

In [115]:
# (total num of non nan female * average female age) + (total num of non nan male * average male age) / total of non nan values for female and male
((61*28.34) + (69*32.99)) / (61+69)

30.808076923076925

# The calculated mean is 30.67. When we remove the NaN values, the answer is 30.81